hudi-client/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -51,6 +51,7 @@
 import org.apache.hudi.common.util.HoodieTimer;
 import org.apache.hudi.common.util.Option;
 import org.apache.hudi.common.util.ValidationUtils;
+import org.apache.hudi.common.util.collection.Pair;
```

Comment:
Can you re-use this same thing to `getAllPartitionPaths()` here: https://github.com/apache/hudi/blob/rfc-15/hudi-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadata.java#L149

The default mechanism is really slow. I investigated `FsUtils.getAllPartitionPaths()` is also used at multiple places across Hudi code-base which can benefit from RFC-15. So, what I am thinking is by default we can let `FsUtils.getAllPartitionPaths()` always default to the metadata table, and internally if the table is not present, it will use this default parallelized/optimized listing of partition paths.

So, it benefits both customers who use metadata table and who don't.

Reply:
>So, what I am thinking is by default we can let FsUtils.getAllPartitionPaths() always default to the metadata table, and internally if the table is not present, it will use this default parallelized/optimized listing of partition paths.

+1 on this. There is a JIRA to track this specifically

Reply:
I wanted to do that but noticed that HoodieBackedTableMetadata is in hudi-common which does not have spark.

Some new interface / refactoring will be required. Can handle this in a different ticket.

Reply:
We can pass in the `HoodieEngineContext` which can also be plain old java if needed. I ll take a closer look in #2351 and how best we can do this. We can scope this PR to just what you just fixed now